chain/network/src/peer_manager/connection.rs

```diff
@@ -155,48 +147,116 @@ impl Connection {
     }
 }
 
-#[derive(Default)]
-pub(crate) struct Pool(ArcSwap<im::HashMap<PeerId, Arc<Connection>>>);
+#[derive(Clone)]
+pub(crate) struct PoolSnapshot {
+    pub me: PeerId,
+    pub ready: im::HashMap<PeerId, Arc<Connection>>,
+    /// Set of started outbound connections, which are not ready yet.
+    pub outbound_handshakes: im::HashSet<PeerId>,
+}
+
+pub(crate) struct OutboundHandshakePermit(PeerId, Pool);
```

Comment:
Not entirely convinced that this is the best design:

* A bunch of outstanding permits keep refrence to the pool, which creates aliasing, and also possibility of reference cycles.
* Permit dropping feels like a big event: this means that we are transitioning to open connection, or dropped connection, or flipping from outbound to inbound. It feels that we want log/metric this cases explicitly. 


My gut feeling would be to go with drop-bomb like design: https://github.com/matklad/drop_bomb


```rust 
pub struct OutstandingHandshakePermit(PeerId);

impl OutstandingHandshakePermit {
    pub fn complete(self, pool: &connection::Pool) {
        pool.outgoing.remove(&self.0);
        std::mem::forget(self)
    }
}

impl Drop for OutstandingHandshakePermit {
    fn drop(&mut self) { log_assert!(false, "forgot about outstanding connection") }
}
```

WDYT? I haven't yet read the code in the depth, so I don't know whether this'll be actually better :)

Reply:
Dropping a permit means that the handshake has been completed or abandoned. All that happens it that some quota is being freed. I wouldn't consider it a big event. We already keep some gauge metrics using guards, and this thing is in the same spirit. I introduced this thing specifically because I had some broken tests while developing https://github.com/near/nearcore/pull/7451/files and realized that tracking all the ways in which the PeerActor can go out of connecting state is a mess.

Reply:
This thing has an implicit drop when PeerActor is dropped, and an explicit drop when PeerActor is transitioning from Connecting to Ready. I suspect that design-wise this whole thing will require further refactor down the road, so I don't claim it is the ultimate solution either. I'm not convinced whether we would gain much by introducing linear types here if the purpose would be to not forget about logging/metric updates - I was convinced to have linearity for future completion (in one of the previous PRs), where it heavily affects semantics, and it didn't work out great out of the box either, unfortunately.

Reply:
Re: reference cycles: in fact a Weak would probably be better for this case, as we are taking a lock for a side effect, not access, WDYT?

Reply:
replaced Arc with Weak ptr in the Permit.